corgi/tasks/pyxis.py

```diff
@@ -0,0 +1,261 @@
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def slow_fetch_pyxis_manifest(
+    oid: str,
+    save_product: bool = True,
+    force_process: bool = False,
+) -> bool:
+    """Fetch a pyxis manifest from pyxis"""
+    logger.info("Fetching pyxis manifest %s", oid)
+
+    # Fetch and parse the SBOM
+    data = get_manifest_data(oid)
+
+    image_id = data["image"]["_id"]
+
+    result = False
+    repositories = data["image"]["repositories"]
+    logger.info("Manifest %s is related to %i repositories", oid, len(repositories))
+    for repository in repositories:
+        result = result or _slow_fetch_pyxis_manifest_for_repository(
+            oid, image_id, repository, data
+        )
+    return result
+
+
+def _slow_fetch_pyxis_manifest_for_repository(
+    oid: str,
+    image_id: str,
+    repository: Mapping[str, Any],
+    manifest: Mapping[str, Any],
+    save_product: bool = True,
+    force_process: bool = False,
+) -> bool:
+
+    component_name = f"{repository['registry']}/{repository['repository']}"
+    logger.info("Processing slow fetch of %s", component_name)
+
+    completion_time = manifest.get("creation_date", "")
+    if completion_time:
+        dt = dateparse.parse_datetime(completion_time.split(".")[0])
+        if dt:
+            completion_dt = make_aware(dt)
+        else:
+            raise ValueError(f"Could not parse completion_time for build {image_id}")
+    else:
+        # This really shouldn't happen
+        raise ValueError(f"No completion_time for build {image_id}")
+
+    softwarebuild, build_created = SoftwareBuild.objects.get_or_create(
+        build_id=image_id,
+        build_type=SoftwareBuild.Type.PYXIS,
+        defaults={
+            "name": component_name,
+            "source": manifest["image"].get("source", ""),
+            # Arbitrary dict can go here
+            "meta_attr": {
+                "image_id": image_id,
+                "manifest_id": oid,
+                "repository": repository,
+                "incompleteness_reasons": manifest["incompleteness_reasons"],
+                "org_id": manifest["org_id"],
+                "creation_date": manifest["creation_date"],
+            },
+        },
+        completion_time=completion_dt,
+    )
+    if build_created:
+        # Create foreign key from Relations to the new SoftwareBuild, where they don't already exist
+        ProductComponentRelation.objects.filter(
+            build_id=image_id, build_type=SoftwareBuild.Type.PYXIS, software_build__isnull=True
+        ).update(software_build=softwarebuild)
+
+    if not force_process and not build_created:
+        # If another task starts while this task is downloading data this can result in processing
+        # the same build twice, let's just bail out here to save cpu
+        logger.warning("SoftwareBuild with build_id %s already existed, not reprocessing", image_id)
+        return False
+
+    component_version = (
```

Comment:
Fix is underway at https://github.com/redhat-appstudio/release-service-utils/pull/58 to populate those digests in pyxis.

Reply:
For the semantic version, I think we don't have it at the moment, but I've filed https://issues.redhat.com/browse/RHTAP-1590 to make sure its in queue.

Reply:
Ok, I added a comment there as it mentioned the concept of streams which I don't fully understand. Perhaps another option to a tag is for us to compare the images in a "stream" using the creation_date, or last_updated_date instread if you don't to construction the tags from that anyway.

Reply:
I see! We can use creation_date for now as a hack in the absence of RHTAP-1590. We'll only be doing single-stream repos for a while (until RHTAP-990 is done) - we just have to remember to return to this and start using that tag instead of creation_date.